### ISSUE TITLE
Google "search history" survives erasing browsing history

### ISSUE THREAD

**Issue by Woodzrul** [mozilla-mobile/focus-android#1306]
Please can you add the below to Firefox Focus.

1. Ability to remove search history. For example, my previous searches appears in the Google search bar when visiting www.google.com.

2. Menu -> Open with Firefox. Can this option only appear to those who have Firefox installed.

Thanks

**Comment by pocmo** [comment]
* Can you provide us with the exact steps to reproduce (1)?

* Your second issue was recently fixed in #1025 and will be released soon.

**Comment by pocmo** [comment]
Also: What Android version / device are you using? :)

**Comment by phunkular** [comment]
I believe that I noticed this (1) before and came to the conclusion that Google just knows it's still you even if you clear history by tapping the trashcan or using the notification... it's kind of creepy actually. I just use DDG now. I think that if you clear focus from memory by swiping, with a little time and/or different IP address, your searches will disappear.

**Comment by Woodzrul** [comment]
Device: OnePlus 2 A2001
Version: 6.0.1

Steps below on reproducing issue 1.

Launch Focus -> Browse to www.google.com -> Search for something, like 'Nintendo Switch'

Despite having cleared the History within Focus and closing down the application my past search results from www.google.com still appear in the drop down menu (search bar). This behavior happens with other sites too I've noticed that use the same searching technology. Helpful for some I'm sure. But thought I should raise it.

See attached screenshot.
![screenshot_20170918-163824 1](https://user-images.githubusercontent.com/23718359/30550724-37c552e0-9c90-11e7-8f7c-7704b7499566.png)

**Comment by npark-mozilla** [comment]
@pocmo I can repro this in sim for API 23. (remove/reinstalling app removes history) we had this issue a long time ago in Focus iOS - fixed by disabling localStorage (I think we already do that) would this be related in this instance too? https://bugzilla.mozilla.org/show_bug.cgi?id=1319208

it's strange, because I remember specifically looking for this a while ago, and didn't see this.
use some non-common search term like 'nojun' :)

**Comment by pocmo** [comment]
We enabled local/session storage to fix a bug in #927. However we wrote some code to clear it whenever a session is erased.

**Comment by Woodzrul** [comment]
Just want to say thank you for your attention on this. Hopefully will be addressed very soon.

**Comment by mcomella** [comment]
To summarize so far:
- I can reproduce on the API 23 emulator. Note that search history will persist even beyond process restarts
- I can reproduce on desktop Chrome using responsive design mode
- Clearing `localStorage` on desktop Chrome will prevent the history from appearing. In theory, one hack fix would be to call `localStorage.clear()` when opening Google and other affected pages but it's unclear whether the page would be able to access (and snoop) these values before we're able to clear manually.
- Using remote dev tools, I found that when we hit the trash can, `localStorage` is not cleared between sessions on the API 23 emulator (using webkit version 44.0.2403.119) and manually clearing it fixes the issue

I'll keep looking for why `WebStorage.deleteAllData` does not appear to remove local storage: either we should be using a different method, it's not possible, or it's a WebKit bug.

**Comment by mcomella** [comment]
In a raw WebView, `WebStorage.deleteAllData` clears `localStorage` which makes me think we're not calling it as it's expected to be called (e.g. the async thing I mentioned).

**Comment by mcomella** [comment]
I commented out [`deleteContentFromKnownLocations` in `cleanup`](https://github.com/mozilla-mobile/focus-android/blob/8940f45ba74e288f27c8301ed058a54211b3c0d8/app/src/webview/java/org/mozilla/focus/webview/SystemWebView.java#L196) and now it works correctly – apparently we've been interfering with WebKit.

One side effect of this is that the first trash can hit successfully clears the search history but subsequent ones do not.

**Comment by mcomella** [comment]
Like the previous, I just removed all invocations of `deleteContentFromKnownLocations`, which deletes files from the WebView directory, and deleting `localStorage` is working correctly. This matches my hypothesis that we're messing with WebKit's delete process.

**Comment by mcomella** [comment]
It appears the problem is with `FileUtils.deleteWebViewDirectory(context);`

**Comment by mcomella** [comment]
I modified the delete web view function so that it will delete everything but the "Local Storage" subdirectory and its contents and localStorage/Google search history is successfully cleared. If I delete the files inside the directory, localStorage/search history is not cleared. Therefore, we have to leave the files inside the "Local Storage" directory on delete. Unfortunately, these files give away domain names (not paths), e.g.:

```
https_news.google.com_0.localstorage
https_news.google.com_0.localstorage-journal
https_support.google.com_0.localstorage
https_support.google.com_0.localstorage-journal
https_www.google.com_0.localstorage
https_www.google.com_0.localstorage-journal
```

So we have to trust WebKit to not leak this data. In theory, we should be able to delete the files on process deletion/creation, which isn't ideal.

---

@pocmo FYI.

Thought: deleting the web view directory prevents the browser from getting rid of local storage – what else could it be preventing? Perhaps this is the cause of #1392 – visited colors persist?

Observation: when we delete the web view directory, WebKit does not restore the directories we delete (i.e. I deleted everything but “Local Storage” when the trash can is hit and LS is the only directory that exists when I load more pages) until the process is recreated.

Hypothesis: when WebKit can’t write state to files (because we deleted them), it refuses to clear its in-memory cache

**Comment by mcomella** [comment]
> Perhaps this is the cause of #1392 – visited colors persist?

Nope: I commented out the delete code and visited colors still persist.

**Comment by mcomella** [comment]
Deleting the WebKit dir was added in #433 (PR #476) but it seems it was a preventative measure rather than one trying to solve a specific problem.

In the worst (and simplest) case, I think we should keep the "Local Storage" folder and its contents when we clear state. Further investigation to continue with preventative measures:
- Deleting "Local Storage" contents when the process starts (in `Application.onCreate`; I confirmed  this works)
- Deleting the "Local Storage" contents after the process ends (since there's no callback, we might have to schedule a repeating job in onStop)
- Tricking WebKit into deleting its Local Storage in memory while deleting the files: e.g. this could entail copying the files into memory while deleting them from disk and restoring them to disk just before we call delete, 
- File a bug with Chromium mentioning that it leaks data through the filesystem despite clearing it in the UX

@pocmo Do you have any thoughts on the best approach?

---

As per https://github.com/mozilla-mobile/focus-android/issues/1306#issuecomment-336207972, we should also consider investigating whether deleting WebKit files on trash can is a good idea.

**Comment by pocmo** [comment]
I wrote WebViewDataTest which does a couple of things and then checks the data directory whether there are any traces of the browsing session. This test showed that WebView leaves traces even when calling all available clear()/delete()/etc. methods.
Debugging another issue I even realized that some things stay around as long as the Chrome process is alive (Although maybe that's a side effect of the issue you are seeing here).

> So we have to trust WebKit to not leak this data. In theory, we should be able to delete the files on process deletion/creation, which isn't ideal.

I would be very cautious with trust here. There are a lot of different versions of WebView/Chromium/Chrome we run on top of across various devices. That's a big drawback of not owning the web engine.

> Thought: deleting the web view directory prevents the browser from getting rid of local storage – 
> what else could it be preventing? Perhaps this is the cause of #1392 – visited colors persist?

The root issue of the link coloring problem is known and tracked by #990 and the linked Chromium bug. Other issues we saw were our workaround not working ("Remove :visited link colors") and other memory/chrome-process caches. Nevertheless it would be interesting to know whether we break more - I think I saw some cookie write log errors from WebView at times.

Overall I'm not happy about our approach of deleting everything. Especially it's now called way too often. When I refactored the app to support multiple tabs I was wondering if we can limit that to very specific points only (e.g. app start to clear whatever survived from the last time). There's also a chance that over-doing it influences other tabs if they kill each others persisted data.

> In the worst (and simplest) case, I think we should keep the "Local Storage" folder and its contents when we clear state

Can you try this and run WebViewDataTest to see if there's still data surviving? This /could/ be problematic: We saw that reddit uses Local Storage to mark subreddits as read. If this data survives then we introduces a new problem. However now I think that was the case where I saw that this data survives our process lifetime and I had to kill the Chrome process (or wait) to kill it. I wonder if removing the unnecessary deletes fixes that?

> Deleting "Local Storage" contents when the process starts (in Application.onCreate; I confirmed this works)

That's what I was wondering too and I think we could try that.

> Deleting the "Local Storage" contents after the process ends (since there's no callback, we might have to schedule a repeating job in onStop)

This sounds hacky. I saw that a Service set to sticky will restart (and also the process!). Maybe we could use that to clean up and then shut it down again. But that's a bit hacky too.

> File a bug with Chromium mentioning that it leaks data through the filesystem despite clearing it in the UX

That's always a good idea. From the other bugs it's very unlikely that they are going to deploy a fix that will change an existing WebView API behavior - even if it's weird. But they mentioned that they are working on some kind of support library that allows users of it to call new APIs that are not exposed otherwise and with that get new features/behaviors (e.g. clear link coloring).

> As per #1306 (comment), we should also consider investigating whether deleting WebKit files on trash can is a good idea.

I agree and think it's actually a pretty bad idea if this happens whenever a tab(!) gets removed. It's not possible via the erase button but via back button for example.

**Comment by pocmo** [comment]
We say we do not store any history etc. - So I think our requirements are:
* Make sure that we do not accumulate traces of browsing history on disk. Temporary caches are okay.
* Do not leak any data to other apps.

I think as long as we stay in those bounds we can be flexible.

**Comment by mcomella** [comment]
> Make sure that we do not accumulate traces of browsing history on disk. Temporary caches are okay.

@pocmo how long do you think temporary data should persist? i.e. is clearing data files on process start-up good enough or should we be trying to do better? Some attack vectors for temporary files (like these localStorage domains) are:
1. Third party APIs, including WebKit, access these files and do something with the information. I'm not sure it matters how long these files are on disk, but just the fact that they're written to disk at all so clear on start-up is probably good enough
1. Users can open the filesystem on unencrypted devices to find these files (fwiw, it looks like [new official Android devices since Marshmallow are encrypted by default](https://nakedsecurity.sophos.com/2015/10/21/new-android-marshmallow-devices-must-have-default-encryption-google-says/))
1. Rooted devices (afaik, i) in some implementations, everything runs as root and apps can read any other apps data and ii) Users can `adb shell` into the device to read files): I assume we don't care about this case

(1), in addition to how WebKit seems to use these files, makes me think it's fine to clear on process creation and (2) makes me think we should additionally have a clean-up job that will run a few hours after Focus is closed (e.g. create the job in `onPause`; this job will have to make sure the Focus process is not still running though so could become complicated).

---

> > In the worst (and simplest) case, I think we should keep the "Local Storage" folder and its contents when we clear state
>
> Can you try this and run WebViewDataTest to see if there's still data surviving?

The test fails because it expects the webview data directory to be deleted and I intentionally delete all the subdirectories but "Local Storage".

> This /could/ be problematic: We saw that reddit uses Local Storage to mark subreddits as read. If this data survives then we introduces a new problem.

By using the dev tools (opening a console on the page and checking `localStorage.length`), I was able to determine `WebStorage.deleteAllData` will clear the localStorage such that the page cannot access the data. However, the files are left on disk: I haven't checked what their contents contain but their filenames expose the domains the user has visited. We should decide on an appropriate solution based on the attack vectors I've listed above (and any you can think of).

> However now I think that was the case where I saw that this data survives our process lifetime and I had to kill the Chrome process (or wait) to kill it. I wonder if removing the unnecessary deletes fixes that?

> Debugging another issue I even realized that some things stay around as long as the Chrome process is alive (Although maybe that's a side effect of the issue you are seeing here).

Yes, deleting the "Local Storage" files appears to block Chrome from deleting the localStorage data exposed to the web page (until the process is killed), which is why I suggest that we should, in the simplest case, leave the files when hitting erase but delete them at some later point.

> > As per #1306 (comment), we should also consider investigating whether deleting WebKit files on trash can is a good idea.
>
> I agree

I filed #1472.

**Comment by mcomella** [comment]
Very curious! In my [RawWebView project](https://github.com/mcomella/RawWebView/tree/webstorage-delete-leaves-files) (basically a web view with a `WebStorage.deleteAllData` button), the local storage files are deleted successfully – I wonder what Focus is doing differently. Maybe we don't need a special solution (beyond don't delete the "Local Storage" dir) after all.

**Comment by mcomella** [comment]
Commenting out our manual deletion of the web view files, `WebStorage.deleteAllData` will *sometimes* remove the local storage files in Focus and always will in my test project (using the same test cases). My test case:
1) Navigate to www.google.com
2) Search "apple", hit enter, wait for page load
3) Clear local storage (trash can, or clear button)

On Focus, sometimes this will clear the local storage files and sometimes this will not. In my testing just now, Focus will always clear the files if we leave out the "apple" step. I wonder if we're hiding/destroying the WebView before we can complete all these async delete calls (assuming they're async). I tried commenting out all the other delete state calls and it doesn't seem to have made a difference.

**Comment by mcomella** [comment]
Even more curious: ^ will actually always delete the localStorage files in Focus but *they get re-added shortly after they're deleted*. I don't know why this isn't the case in my test app.

**Comment by mcomella** [comment]
If I comment out [the line that removes the current session](https://github.com/mozilla-mobile/focus-android/blob/9210c6a339a1c4ec9733eb1e2becbb1daa546a11/app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java#L702) when erase is hit, the localStorage files are not restored. I wonder if this is a problem with how we, or the system, stores the WebView state or the fragment state when it's replaced.

**Comment by mcomella** [comment]
Additionally, this only happens if the google search results page is actively open, i.e. if I navigate from google.com to search "apple" and click on apple.com, then erase, all Local Storage will be erased. However, if the search results are left open, the google.com entries will be restored (note: non-google entries will not be restored).

I wonder if we're cleaning up before we close the web view so google.com saves its state again after we've cleared state.

**Comment by pocmo** [comment]
> @pocmo how long do you think temporary data should persist? i.e. is clearing data files on process start-up good enough or should we be trying to do better?

I think if we can't do better then having temporary data until the next app start is okay. However we should not have it over multiple app starts. All this upon condition that this data does not interfere with  the browsing session (e.g. users are still logged in etc.).

> Very curious! In my RawWebView project (basically a web view with a WebStorage.deleteAllData button), the local storage files are deleted successfully – I wonder what Focus is doing differently. 

This reminds me of a thing we do and you could try disabling it (random guess though): Whenever the WebFragment is paused we save the current WebView state in the Session object. This is to make sure that we can restore this WebView later - for example when switching tabs or the UI components get destroyed while the app is in the background. Of course we run through WebFragment.onPause() too when the whole thing gets destroyed. This is usually not a problem because we are throwing the Session object away too. However it might be possible that the WebView will also write data to disk when we save the state (in a Bundle). Not sure if it's related but this is one thing we might do differently than in your test project. :)

Overall this is yet another advertisement for using Gecko - a true private mode doesn't save data in the first place. Too bad they deprecated this API in WebView.

**Comment by mcomella** [comment]
> Whenever the WebFragment is paused we save the current WebView state in the Session object.

Commenting this out doesn't seem to make a difference.

---

I stepped through the debugger and discovered WebKit writes the state just after `WebFragment.onPause` returns, specifically [when the WebView loses focus](http://androidxref.com/6.0.1_r10/xref/frameworks/base/core/java/android/webkit/WebView.java#2521) as a result of [the FragmentManager removing the view](http://androidxref.com/6.0.1_r10/xref/frameworks/support/v4/java/android/support/v4/app/FragmentManager.java#1126).

Android Studio doesn't seem to have the WebView source code so I can't use debugger so it'll be hard to figure out why specifically this is happening.

**Comment by mcomella** [comment]
The debugger still works, without source, so I was able to keep a copy of the source (at perhaps not the correct revision) in the browser and step through it. The state seems to get saved in ContentViewCore.onFocusChanged:

```java
if (mNativeContentViewCore != 0) nativeSetFocus(mNativeContentViewCore, gainFocus);
```

To native code, where I definitely can't step through. :)

I think the ideal solution is to figure out the last time this is called (I assume it's only called here) and find a later callback that we can use to call `cleanup`.

**Comment by mcomella** [comment]
I added a cleanup after the focus change and it deleted the localStorage, but then localStorage the localStorage is restored again later. I even tried to cleanup in `UrlInputFragment.onResume` but it restores the localStorage data (onDestroy in the old fragment gets called after onResume in the new one - maybe that's related).

It seems like it's going to be non-trivial to find a callback after which the WebView won't restore the data, especially because the restoration can be happening asynchronously.

**Comment by mcomella** [comment]
Ideally, we delete LocalStorage in response to erase or the fragment being removed or stopped/destroyed so we don't have to pass state to other places in the application. Unfortunately, all the obvious callbacks (onDetach, onDestroy, onPause) don't seem to work because the data is restored. Two callbacks that work are `UrlInputFragment.onResume` and `UrlinputFragment.onFilter` - these are not ideal because we'd have to transmit the "trash can was hit" state to other parts of the application, which adds complexity and is liable to add bugs.

I'm looking at an alternative – disabling the WebView before we delete the data – so that we can erase the data in the BrowserFragment. We might be able to disable the WebView by loading a dummy page, removing it from the View hierarchy, etc.

Note: LocalStorage is only removed when the method that after `WebStorage.deleteAllData()` returns so that call must only queue the event (unclear if it's async or serial).

**Comment by mcomella** [comment]
> Note: LocalStorage is only removed when the method that after WebStorage.deleteAllData() returns so that call must only queue the event (unclear if it's async or serial).

Looks like [it gets queued to run on the main thread](https://cs.chromium.org/chromium/src/android_webview/browser/aw_quota_manager_bridge.cc?type=cs&q=deletealldata&sq=package:chromium&l=198), although we shouldn't really on that behavior.

**Comment by mcomella** [comment]
> Unfortunately, all the obvious callbacks (onDetach, onDestroy, onPause) don't seem to work because the data is restored.

I also tried `FragmentTransaction.runOnCommit` and that also didn't work.

One thing that works is loading `about:blank` and, in `onPageFinished` we call `cleanup` and clear the session. It's fragile (e.g. what happens when we erase from the notification shade?) but it doesn't force the responsibility onto a separate Fragment. I think this might be the best we can do.

I tried removing WebView from the view hierarchy but this didn't seem to work.

**Comment by mcomella** [comment]
To take a step back, our goal is to make sure LocalStorage (and thus Google search history) does not persist to the next page load after you click erase.

So far, two solutions stand out:
1. When erase is requested, load `about:blank`, wait for it to load, clear the web view data, and remove the current session (which will return to the `UrlInputFragment` via LiveData observers). This is incomplete because it does not handle the case that erase is requested in the background so we'll probably also need to...
1. Erase the webView state when starting a new browsing session. This is not ideal because:
    1. it's detached from the cause of the erase (that the session was erased), creating complexity in identifying the states in which we should clear this history and being able to maintain this code over time.
    1. For one problem site per session, such as the google search results, it would allow LocalStorage files, that give away domain names, to stay on disk until the next browsing session begins. This would be vulnerable to local attack vectors (https://github.com/mozilla-mobile/focus-android/issues/1306#issuecomment-336301381) but would not leak data through the browser.
    1. The callbacks that restore the LocalStorage files are queued or async, so it might be impossible to find an intuitive callback that'd guarantee that we can delete LocalStorage and it won't be restored shortly afterwards. e.g. `UrlInputFragment.onFilter` and `BrowserFragment.onCreateViewCalled` are the only callbacks I've found that work so far

Current thinking: it seems like Sebastian would think (2) is good enough and it'd be simpler to keep the changes in one place so I'll see if it's possible to get it to work 100% of the time (see 2.iii).

**Comment by mcomella** [comment]
Here's what iOS does:
- [`BrowserViewController.resetBrowser`](https://github.com/mozilla-mobile/focus-ios/blob/083d98a759f03817c14776f080ada68ede6cd7bf/Blockzilla/BrowserViewController.swift#L212) resets the WebView, calling two different `reset` methods:
    - [`WebCacheUtils.reset`](https://github.com/mozilla-mobile/focus-ios/blob/e8112be19bc3ddee04777b421496876455e5300d/Blockzilla/WebCacheUtils.swift#L8) which removes static WebView caches/storage
    - [`WebViewController.reset`](https://github.com/mozilla-mobile/focus-ios/blob/083d98a759f03817c14776f080ada68ede6cd7bf/Blockzilla/Modules/WebView/WebViewController.swift#L55) which loads "about:blank" in the current webview, cleans up its resources, and replaces it with a new browser
- (per above) iOS loads about:blank but there's no great explanation why. It was implemented in [bug 1319297](https://bugzilla.mozilla.org/show_bug.cgi?id=1319297), which mentions, "Turns out the specific cache was the in-memory back/forward cache. From some experimentation, I found that waiting a short period before recreating the UIWebView fixes this.". The about:blank change was added without explanation, however.

**Comment by mcomella** [comment]
I finished the bug so I removed the breakdown tag.
